### ISSUE TITLE
Option for symmetrizing in Tensor constructors

### ISSUE THREAD

**Issue by TakeTwiceDailey** [KeitaNakamura/Tensorial.jl#198]
I think it would be helpful to add an option in the construction for symmetric tensors to apply a symmetrizer to the indices that are meant to be symmetric.

Suppose I want to construct a `SymmetricFourthOrderTensor` from four vectors like so:

$Q_{ijkl}=A_{(i}B_{j)}C_{(k}D_{l)}$ 

I might naively try
```
Q = SymmetricFourthOrderTensor{4,Float64}((i,j,k,l) -> A[i]*B[j]*C[k]*D[l])
```
But the constructor here does not symmetrize the vectors, it simply picks out the components on the lower triangle of the indices $(i,j)$ and $(k,l)$. I could write out the symmetric operation by hand, but this is annoying and prone to mistakes, especially when the constructor is much more complicated than the above example. It would be nice to give an argument meant to tell the constructor to symmetrize the relevant indices. Is there already an ideal way to do this that I am unaware of?

**Comment by KeitaNakamura** [comment]
Hi. Currently, there is no function for performing arbitrary symmetrization. Could you provide the equation for this?

**Comment by TakeTwiceDailey** [comment]
Just like you have implemented for second rank tensors

$Q_{ij} = A_{(i}B_{j)} = \frac{1}{2}(A_i B_j + A_j B_i)$

But now each symmetric pair is done separately

$Q_{ijkl} = A_{(i}B_{j)}C_{(k}D_{l)} = \frac{1}{4}(A_i B_j + A_j B_i)(C_k D_l + C_l D_k)$

Technically there is also a procedure to take the symmetric part of an arbitrary number of indices ([https://www.wikiwand.com/en/Symmetric_tensor#Symmetric_part_of_a_tensor](url)), but I don't that would be that useful here.

While we are on this topic, it might be nice to implement anti-symmetric tensors

$P_{ij} = A_{[i}B_{j]} = \frac{1}{2}(A_i B_j - A_j B_i)$

as well as operations for anti-symmetrization.

**Comment by KeitaNakamura** [comment]
If you just want the specific `symmetric` function to convert `FourthOrderTensor` into `SymmetricFourthOrderTensor`, PR is welcome (it's easy to implement as you wrote the equation). Maybe the name should be `minorsymmetric` rather than `symmetric`.

**Comment by KeitaNakamura** [comment]
For anti-symmetrization, the `skew` function is already implemented. I don't have a strong motivation to implement skew-symmetric tensors. Using symmetric tensors is primarily for reducing duplicated computations to improve performance. I think there aren't many cases where we need the speed to calculate skew-symmetric tensors than symmetric tensors.

**Comment by TakeTwiceDailey** [comment]
An operation like `symmetric` to convert `FourthOrderTensor` into `SymmetricFourthOrderTensor` is one part of this. However, as a user, I don't want to have to make a `FourthOrderTensor` and then reduce that to a  `SymmetricFourthOrderTensor`, because that seems like a waste of computation and memory (maybe it isn't?). Instead, it would be nice to have the construction mechanism for `SymmetricFourthOrderTensor` to know how to do this at the construction level, so no `FourthOrderTensor` is necessary. 

I have to admit I am not familiar with the kind of Julia machinery that the code in this package uses. I am still fairly new to this language. I guess the `symmetric` function would look like this:
```
@inline symmetric(x::AbstractFourthOrderTensor{dim}) where {dim} =
    SymmetricFourthOrderTensor{dim}((i,j,l,k) -> (x[i,j,k,l]+x[j,i,k,l]+x[i,j,l,k]+x[j,i,l,k])/4)
```
But for this to happen at the construction level, I am afraid I don't know what that would look like.

**Comment by TakeTwiceDailey** [comment]
I guess what would happen under the hood is when a tensor is being constructed, I could supply a keyword argument (or something) like `symmetrize=true`, and then the function I supply
```
    SymmetricFourthOrderTensor{dim}(f(i,j,k,l),symmetrize=true)
```
gets replaced with
```
    SymmetricFourthOrderTensor{dim}((i,j,k,l)->(f(i,j,k,l)+f(j,i,k,l)+f(i,j,l,k)+f(j,i,l,k))/4)
```

**Comment by KeitaNakamura** [comment]
If you want the new constructor, I don’t want to implement it unless arbitrary symmetrization can be supported (I don’t want to define the special constructor only for the SymmetricFourthOrderTensor.)

**Comment by KeitaNakamura** [comment]
If you're concerned about performance, please compare and report it first. Allocation doesn't occur since the tensor is of the bits type.

**Comment by KeitaNakamura** [comment]
Closed by #199.

```jl
julia> A, B, C, D = reinterpret(Vec{3,Float64}, rand(12));

julia> Q = A ⊗ B ⊗ C ⊗ D;

julia> minorsymmetric(Q) ≈ @einsum (i,j,k,l) -> (Q[i,j,k,l] + Q[j,i,k,l] + Q[i,j,l,k] + Q[j,i,l,k]) / 4
true
```
